3-Register/index.html

```diff
@@ -0,0 +1,235 @@
+                <path d="M49.2339 39.5595C49.775 39.0184 50.6523 39.0184 51.1935 39.5595C51.7346 40.1006 51.7346 40.978 51.1935 41.5191C50.6523 42.0602 49.775 42.0602 49.2339 41.5191C48.6927 40.978 48.6927 40.1006 49.2339 39.5595Z" fill="#F87171"/>
+                <path d="M47.168 25.1654C46.6269 24.6243 45.7496 24.6243 45.2084 25.1654C44.6673 25.7065 44.6673 26.5839 45.2084 27.125C45.7496 27.6661 46.6269 27.6661 47.168 27.125C47.7092 26.5839 47.7092 25.7065 47.168 25.1654Z" fill="#F87171"/>
+            </svg>
+            
+            <h3 id="sign-in-msg" class="sign-in-text"></h3>
+        </div>
+    </div>
+    <script>
+        let passwordVisible = false
+        let confirmPasswordVisible = false;
+
+        const emailIsValid = function (mail) {
+            const availableDomains = ['@gmail.com', '@hotmail.com', '@lightit.io']
+            const mailHasAcceptedDomain = availableDomains.some(domain => mail.endsWith(domain))
+
+            if (! mailHasAcceptedDomain) {
+                return populateErrors({ email: 'Email ends with invalid domain' })
+            }
+
+            return true
+        }
+
+        const usernameIsValid = function (username) {
+            const allowedChars = /^[a-z]+$/
+
+            if (username.length < 3) {
+                return populateErrors({ username: 'Usernames must have 3 or more characters' })
+            }
+
+            if (!allowedChars.test(username)) {
+                return populateErrors({ username: 'Usernames can only contain lowercase characters' })
+            }
+
+            return true
+        }
+
+        const passwordIsValid = function (password) {
+            const correctLength = password.length >= 8
+            const hasUppercase = /[A-Z]/.test(password)
+            const hasLowecase = /[a-z]/.test(password)
+            const hasSpecialChar = /[!@#$%^&*()_+{}[\]:;<>,.?~\\/-]/.test(password)
+
+            if (!correctLength) {
+                return populateErrors({ password: 'Password must be 8 characters or more'})
+            }
+
+            if (!hasUppercase) {
+                return populateErrors({ password: 'Password must have at least one uppercase character'})
+            }
+
+            if (!hasLowecase) {
+                return populateErrors({ password: 'Password must have at least one lowercase character'})
+            }
+
+            if (!hasSpecialChar) {
+                return populateErrors({ password: 'Password must have at least one special character'})
+            }
+
+            return true
+        }
+
+        const passwordAndConfirmMatch = function () {
+            const pswd = document.querySelector('#password-input')
+            const confirm = document.querySelector('#password-confirm-input')
+
+            if (pswd.value !== confirm.value) {
+                return populateErrors({ confirm: 'Password & Confirm Password fields do not match' })
+            }
```

Comment:
When you read the name `passwordAndConfirmMatch`, you wouldn't imagine that this function is going to modify the error fields.

I think this kind of side effect in a function is not a good practice, I would just return `pswd.value !== confirm.value` and then populate the error in the `validateForm` function.

To put it in another way, the responsibility of setting the error fields should be assigned to the `validateForm` function, not the `passwordAndConfirmMatch` one, since the latter one should only focus on verifying whether these two passwords match or not.

Reply:
Nice. I agree that these functions (`passwordAndConfirmMatch` and the previous ones as well) include side effects which are not clear by the name of the function. One benefit I saw of handling it like this was was that, in `validateForm`, the snippet

`const isValid = emailIsValid(email) &&
                usernameIsValid(username) &&
                passwordIsValid(password) &&
                passwordAndConfirmMatch()`
                
handled the validation logic very concisely (if one validation failed the other ones weren't executed and the corresponding errors were displayed).

If I were to remove the side effects from `passwordAndConfirmMatch` and the other functions, what would be the best way to handle this? Perhaps the functions in charge of checking if a field is valid can return an object like `{valid: false, error: ...}`, and `validateForm` can then populate the errors. Is there a way to handle this logic that isn't too verbose (without a bunch of `if` conditionals inside `validateForm`)?

Reply:
What you suggested sounds great!